### PR TITLE
FRC46 factory token and token invoke() helper

### PR DESCRIPTION
**Pull request by abright**
Implements a token suitable for #128 where the code can be deployed once and multiple tokens created by changing parameters passed into the `Constructor` call. The token name, symbol, and granularity are set at construction time.

Minting is controlled by a single authorised user which is also set at construction time - this address is optional because it's going to be paired with an initial supply (so one or the other must be set and will be checked during construction) to cover the common faucet and fixed supply strategies. A new method `DisableMint` is added that will clear the authorised user to disable minting.

Also introduces `frc46_invoke` which is a generic function to handle dispatch of all standard FRC-0046 token methods (see #127)
This invoke helper function should move out of the token implementation into a library, but it needs a bit more work around error handling so I'll follow up on that later in a new PR.

### CODE REVIEW

File: token_examples/factory_token/src/token.rs
Comment (abright) [inline token_examples/factory_token/src/token.rs:112]: Probably need a separate read-only implementation of the `Token` handle to clean this up (also for `total_supply` and any future query/getter functions)

File: token_examples/factory_token/src/token.rs
Comment (abright) [inline token_examples/factory_token/src/token.rs:308]: Thoughts? I did what was easy (the single operator for minting) but not sure if it's the best approach

**Review by anorth — COMMENTED** [review]
Thanks, this looks directionally right. Biggest thing is to fix the APIs to make things like the token trait and library work together, and enable the generic dispatch thing. 

I didn't look at the tests too much  yet. Is there room to reduce testing overheads by using this actor for the main library integration tests, or is the test actor there too specialised? We do need to refactor to reduce the duplicated code tho.

File: token_examples/factory_token/src/token.rs
Comment (abright) [reply token_examples/factory_token/src/token.rs:96]: Generic invoke needs the trait, but maybe it's worth pulling out of where it lives now, reworking it, and keeping it together with the invoke function in some other place - I feel like that stuff belongs in a utility module or library separate to the core token library code. 

I've made the error an associated type which is a bit nicer to live with and doesn't affect existing code too much, still need to do some work on the read-only calls like `total_supply()` though. I'm leaning towards just making them take `&mut self` like most of the other functions already do

File: token_examples/factory_token/src/lib.rs
Comment (abright) [reply token_examples/factory_token/src/lib.rs]: It was a leftover from a plan to have a minter and/or an initial supply amount. We dropped the initial supply part of it so I've removed the option from constructor params now too.

File: token_examples/factory_token/src/token.rs
Comment (abright) [reply token_examples/factory_token/src/token.rs]: It's now a single `ActorID`, as suggested in another comment

File: token_examples/factory_token/src/token.rs
Comment (abright) [reply token_examples/factory_token/src/token.rs]: Done. I made the return data serialization abort on errors to avoid the minefield of it being compatible with whatever error type a user implements. It should be in a state where it could be used in external code now.

File: token_examples/factory_token/src/token.rs
Comment (abright) [reply token_examples/factory_token/src/token.rs]: Took a bit of crate trickery to set up but the main battery of integration tests now runs and passes using this token

File: token_examples/factory_token/src/lib.rs
Comment (abright) [reply token_examples/factory_token/src/lib.rs]: I've changed it to have invoke() abort on error results and moved the method dispatching into a new function so it can return results. Clippy is mad about the size of the error type but that doesn't seem to be related to changes in this PR - I see it happening for https://github.com/helix-onchain/filecoin/pull/166 as well. I'll address that in a separate branch.

File: token_examples/factory_token/src/token.rs
Comment (abright) [reply token_examples/factory_token/src/token.rs]: Addressed in https://github.com/helix-onchain/filecoin/pull/149#discussion_r1028709951 - I've changed the underlying trait to take mutable refs for these functions

**Comment by abright** [comment]
This now passes our integration tests when run instead of the basic_token_actor. I think we can remove that older token actor now, or maybe move it elsewhere to keep as an example?

I should probably move this one too - drop the `token_example` directory and make it a top level thing, maybe with a different name like `frc46_token_factory` ?

File: token_examples/factory_token/src/lib.rs
Comment (anorth) [inline token_examples/factory_token/src/lib.rs:72]: It looks like this will overwrite any exit code provided by the token library when frc46_invoke returns an Err, e.g. insufficient balance

File: token_examples/factory_token/tests/common.rs
Comment (anorth) [inline token_examples/factory_token/tests/common.rs:141]: This is actually not that high. Can you import BLOCK_GAS_LIMIT to here from fvm_shared?

File: token_examples/factory_token/tests/factory_token.rs
Comment (anorth) [inline token_examples/factory_token/tests/factory_token.rs:183]: Please assert the actual exit code expected

**Review by anorth — COMMENTED** [review]
We might want to rename the factory_token to frc46_factory_token, since there'll probably be an NFT one later too

**Review by anorth — APPROVED** [review]
Looks pretty good, but please get review from @alexytsu before merging, especially to discuss traits, dependencies etc.

File: token_examples/factory_token/src/lib.rs
Comment (alexytsu) [inline token_examples/factory_token/src/lib.rs:39]: I'm not quite sure what this means or why it's necessary

File: token_examples/factory_token/token_impl/Cargo.toml
Comment (alexytsu) [inline token_examples/factory_token/token_impl/Cargo.toml:15]: as noted above - 2.0.0 is stable

**Review by alexytsu — APPROVED** [review]

File: token_examples/factory_token/src/lib.rs
Comment (abright) [reply token_examples/factory_token/src/lib.rs:72]: Added exit codes to the `RuntimeError` type so it can return the proper code here

File: token_examples/factory_token/tests/factory_token.rs
Comment (abright) [reply token_examples/factory_token/tests/factory_token.rs:183]: Done, now that it does return a meaningful exit code for each error

File: token_examples/factory_token/tests/common.rs
Comment (abright) [reply token_examples/factory_token/tests/common.rs:141]: Did this for the main integration tests too

**Pull request merged**
